Added hourly backup option. Added exclude file. Modified script sligh…

jhuscott opened the pull request:
…tly to more closely mimic the behavior of the Borgbase Vorta backup client. Added more documentation.

jhuscott committed:
Added hourly backup option. Added exclude file. Modified script slightly to more closely mimic the behavior of the Borgbase Vorta backup client. Added more documentation.

jhuscott committed:
Added support for BORG_PASSCOMMAND in example.env

bebehei commented on backup:
```diff
@@ -30,7 +31,8 @@ case "$1" in
 			${OPT_BORG} ${OPT_BORG_CREATE} \
 			${EXCLUDE_FILE:+--exclude-from=${EXCLUDE_FILE}} \
-			::'{hostname}'${SUFFIX}'-{now:%Y-%m-%d_%H:%M}' \
+			--exclude-if-present .nobackup \
```
What about putting this into `example.exclude`? I'm a big fan of not hardcoding such things.

jhuscott replied:
That's fair; I'll add a variable that pulls from example.exclude.

jhuscott replied:
So, this will have to be a separate file as it cannot be contained within the EXCLUDE_FILE variable. I'll make the change and update the pull request.

-exclude-if-present documentation says "exclude directories that are tagged by containing a filesystem object with the given NAME". For what I included in the script, if '.nobackup' is present in a directory, it will exclude all files in that directory and subdirectories.

jhuscott replied:
Disregard; I'll make this a variable in example.env. I can't imagine this having more than a couple of filenames for this type of tag, and a variable makes sense rather than coding and splitting this to a separate configuration file (for now).

jhuscott replied:
Done

bebehei commented on backup:
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash
+export PATH=$PATH:/usr/local/bin
```
There's no necessity to add `/usr/local/bin` to the `PATH`. No program is called in this script by its name, *except* `borg`. So adding a `CMD_BORG` variable would be a proper solution and even a great feature!

bebehei replied:
Please get used to call scripts by its command name and not by its PATH. Sometimes, there are good reasons to use full paths, but most of the time it's an antipattern.

jhuscott replied:
That's fair. If that's the case, though, the Makefile needs to be re-written as well because that's where the executables reside.

jhuscott replied:
I should mention that NOT putting the path in the script will cause RHEL6+ and Debian Wheezy+ to maybe not find the Borg executable on a system that's been locked down. Scripts executed by cron/anacron don't end up having a pre-defined PATH variable and the backup script will fail.

bebehei commented on README.md:
````diff
@@ -21,12 +21,19 @@ backup (do)
 ```
 
-Create a new archive and then prune the repository archives. (Wraps `do-create` and `do-prune`)
+Create a new archive and then prune the repository archives. (Wraps `do-create` and `do-prune`. This can also be accomplished by simply invoking `backup`, e.g.: `/usr/local/bin/backup`.)
````
`/usr/local/bin` ... see my comments below.

bebehei commented on README.md:
```diff
@@ -36,14 +43,21 @@ You need to have [borgbackup](https://github.com/borgbackup/borg/) prior to your
 
     # add cronjob (for hourly backup)
-    { crontab -l ; echo "0 * * * * backup"; } | crontab -
+    { crontab -l ; echo "0 * * * * /usr/local/bin/backup"; } | crontab -
```
`/usr/local/bin` ... see my comments below.

jhuscott replied:
When writing a crontab (especially root's crontab), hard paths should be coded for whatever you are running. Not doing so easily allows an attacker to compromise a system. If two "backup" scripts, for example, are on the path, an attacker could write one that would get executed by root and cron would have no way of reporting that.

bebehei replied:
I see your point. But whenever you've got a path in your PATH variable, which is not strictly limited to be writable by the user itself or root, you've got a problem. A different problem and a problem not limited to cron.

The correct solution for that is not to hardcode a single executable path, **you have to hardcode *every* path!**

jhuscott replied:
I updated the README with additional information. I think in this case, providing an example for the crontab entry is OK.

bebehei reviewed (commented):
Are you able to split your commits by topic? It would be a great thing!

bebehei commented:
Also I'd like to apologise for the last months not reacting to anything. I had a personal breakdown and had to take a hiatus.

Commit:
Added CMD_BORG variable

Added CMD_BORG variable to example.env and modified backup script to accomodate.

Added information to README.

Commit:
Moved borg command verification

Moved borg command verification to *after* the environment file is read-in.

bebehei commented on backup:
```diff
@@ -23,22 +20,27 @@ fi
 
 . "${CONFIG_BASE}/${BACKUP_CONF}.env" || die "Could not read envfile."
 
+command -v $CMD_BORG &>/dev/null \
```
```suggestion
: "${CMD_BORG:=borg}"
command -v "${CMD_BORG}" &>/dev/null \
        || die "Cannot find '${CMD_BORG}' executable! Verify the CMD_BORG variable."
```

jhuscott committed:
Update backup

Corrected CMD_BORG variable definition

Co-Authored-By: Benedikt Heine <bebe@bebehei.de>

Commit:
Modify "exclude-if-present"

Created variable called OPT_BORG_EXCLUDE_IF_PRESENT in example.env and added this to the backup script.

Commit:
Update example.exclude

Added documentation link and additional sample exclusion patterns that are useful.

jhuscott committed:
Corrected CMD_BORG check - removed extra conditional

jhuscott commented:
Let me know if I need to make any more changes for the PR. Have a good weekend!

jhuscott committed:
Add BORG_RSH variable

Added BORG_RSH variable to set the location of an SSH key for a remote borg server

jhuscott committed:
Add BORG_REPO examples

Added examples for BORG_REPO variable to include local and remote hosts

jhuscott committed:
--prefix flag deprecated; replaced with --glob-archives

jhuscott committed:
Add compact command, since prune no longer compacts the archive

jhuscott committed:
Add BORG_FILES_CACHE_TTL option in example.env

jhuscott committed:
Expand verbose output options and add the progress and stats to OPT_BORG_PRUNE

jhuscott committed:
Changed default to always retain the last remaining archive.

jhuscott committed:
Verbiage cleanup

jhuscott committed:
Added missing semicolons.  Script now executes.

jhuscott committed:
Added do-compact operation to run default.

jhuscott committed:
Added showBackupInfo script

jhuscott committed:
Updated file exclusions

jhuscott committed:
Updated README

jhuscott committed:
Added borg keyfile options

jhuscott committed:
Updated README to include keyfile information

jhuscott committed:
Added OPT_BORG_EXCLUDE_CACHES

jhuscott committed:
Added warning to back up the borg key upon init

jhuscott committed:
added key backup examples

jhuscott committed:
added blank line after warning

jhuscott committed:
formatting

jhuscott committed:
Formatting

jhuscott committed:
Included info on borg compact, and reformatted README

jhuscott committed:
Added instruction to add a trailing space to borg option variables

jhuscott committed:
Correct typos and reformat README

jhuscott committed:
Excluded duplicate output from showBackupInfo

jhuscott committed:
Corrected do-prune glob-archives option to include wildcard

bebehei commented:
Do you have any intent to get this merged?
